### PR TITLE
[PLGN-621] Add OAuth to ServiceNow Plugin

### PR DESCRIPTION
**Pull request by mrinehart-r7**
### Description

Describe the proposed changes:

  - Adds OAuth support for API authentication by way of adding two new optional fields to the connection: `oauth client ID` and `oauth client secret`

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/servicenow/help.md [hunk 1]
```diff
@@ -1,45 +1,43 @@
 # Description
 
-[ServiceNow](https://www.servicenow.com/) is a tool for managing incidents and configuration management. This plugin allows users to manage all aspects of incidents including creation, search, and updates. Additionally, incident changes can be monitored and processed for use in a Rapid7 InsightConnect workflow.
-
-Note: This plugin affects only the underlying tables in a ServiceNow instance, not its UI. Hence, this plugin will work seamlessly with Virtual Task Boards.
+ServiceNow is a tool for managing incidents and configuration management. Using the ServiceNow plugin for Rapid7 InsightConnect, users can manage all aspects of incidents including creation, search, updates, as well as monitor them for changes
```
Comment (igorski-r7) [inline plugins/servicenow/help.md:3]: Could you please revert description, key features and requirements? 😄

File: plugins/servicenow/help.md [hunk 2]
```diff
@@ -1045,20 +587,158 @@ Example output:
...
+
+|Name|Type|Required|Description|Example|
+| :--- | :--- | :--- | :--- | :--- |
+|security_incident|security_incident|False|Details of the security incident|{}|
+  
+Example output:
+
+```
+{
+  "security_incident": {}
```
Comment (igorski-r7) [inline plugins/servicenow/help.md:683]: What about this example?

File: plugins/servicenow/plugin.spec.yaml [hunk 3]
```diff
@@ -4,7 +4,7 @@ products: ["insightconnect"]
 name: servicenow
 title: ServiceNow
 description: ServiceNow is a tool for managing incidents and configuration management. Using the ServiceNow plugin for Rapid7 InsightConnect, users can manage all aspects of incidents including creation, search, updates, as well as monitor them for changes
-version: 7.3.1
+version: 7.4.0
 connection_version: 7
 supported_versions: ["2020-03-11 Orlando"]
```
Comment (dlaverty-r7) [inline plugins/servicenow/plugin.spec.yaml:9]: Can you confirm if this was the version you completed your testing on?

File: plugins/servicenow/icon_servicenow/util/request_helper.py [hunk 4]
```diff
@@ -78,3 +127,35 @@ def get_attachment(connection, sys_id):
...
+                "username": self.username,
+                "password": self.password,
+            },
+            timeout=30,
+        )
+
+        if response.status_code != 200:
+            error_message = "unknown"
+            if response.status_code == 401:
+                error_message = response.json().get("error_description", "unknown")
```
Comment (dlaverty-r7) [inline plugins/servicenow/icon_servicenow/util/request_helper.py:147]: (Suggestion): Would this be better defaulting to unauthorized for a 401 error?

File: plugins/servicenow/plugin.spec.yaml (hunk 3 above)
Comment (mrinehart-r7) [reply plugins/servicenow/plugin.spec.yaml:9]: Hey Dympna - I have been using New York for testing.

**Review by joneill-r7 — APPROVED** [review]

**Review by igorski-r7 — APPROVED** [review]

**Pull request merged**